Frameworks/Foundation/NSInvocation.mm

```diff
@@ -14,71 +14,99 @@
 //
 //******************************************************************************
 
-#import "Starboard.h"
-#import "Foundation/NSInvocation.h"
+#import <Starboard.h>
+#import <Foundation/NSInvocation.h>
 #import <ctype.h>
-#import "LoggingNative.h"
 #import <objc/encoding.h>
+#import "LoggingNative.h"
 #import "ErrorHandling.h"
 
+#import "NSMethodSignatureInternal.h"
+#import "NSInvocationInternal.h"
+
+#include <vector>
+#include <memory>
+
 static const wchar_t* TAG = L"NSInvocation";
 
-@implementation NSInvocation
-static void* copyArgument(NSInvocation* self, void* buf, int index) {
-    char* type = (char*)[self->_methodSignature getArgumentTypeAtIndex:index];
+@implementation NSInvocation {
+    NSMethodSignature* _methodSignature;
 
-    int size = objc_sizeof_type(type);
+    std::unique_ptr<_NSInvocationCallFrame> _callFrame;
 
-    void* ret = (void*)IwMalloc(size);
-    memcpy(ret, buf, size);
+    size_t _returnLength;
+    BOOL _retainArguments;
 
-    return ret;
+    uint8_t _smallReturnValueOptimization[16];
```

Comment:
Always `memcpy`'d in and out bytewise; alignment should not matter here. It looks like for struct returns the compiler emits a `memcpy` before the epilogue as well.